src/test/java/uk/ac/sanger/sccp/stan/service/TestSlotCopyService.java

```diff
@@ -133,8 +133,27 @@ public void testPerform(boolean valid) {
             assertSame(ex, assertThrows(ValidationException.class, () -> service.perform(user, request)));
         }
         verify(service).performInsideTransaction(user, request);
-        // Unstore will be executed because the optype of the operation has the "discard sources" flag
-        verify(service, times(valid ? 1 : 0)).unstoreSources(user, request);
+        // Unstore will not be executed because the optype does not have the the "discard sources" flag
+        verify(service, never()).unstoreSources(any(), any());
+    }
+
+    @Test
+    public void testPerformAndDiscard() {
+        SlotCopyRequest request = new SlotCopyRequest("op", "thing", List.of(), null);
+        OperationType discardingOpType = EntityFactory.makeOperationType("dot", null, OperationTypeFlag.DISCARD_SOURCE);
+        Operation op = new Operation(200, discardingOpType, null, null, null);
+        OperationResult result = new OperationResult(List.of(op), List.of());
+        when(transactor.transact(any(), any())).then(invocation -> {
+            Supplier<?> sup = invocation.getArgument(1);
+            return sup.get();
+        });
+        doReturn(result).when(service).performInsideTransaction(any(), any());
+        doNothing().when(service).unstoreSources(any(), any());
+
+        assertSame(result, service.perform(user, request));
+        verify(service).performInsideTransaction(user, request);
+        // Unstore will be executed because the optype has the "discard sources" flag
+        verify(service).unstoreSources(user, request);
     }
```

Comment:
Would you please explain what this test does? I couldn't figure it out.

Reply:
The logic in the service now says "if the operation type has the discard-sources flag, then discard the sources; if it has the mark-used flag, then mark the sources used". This method tests that the behaviour would still be correct for an operation type that has the discard-sources flag.